security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,818 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chiron
+
+import (
+	"bytes"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/howeyc/fsnotify"
+	cert "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	certclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/spiffe"
+	istioutil "istio.io/istio/pkg/util"
+	crl "istio.io/istio/security/pkg/k8s/controller"
+	"istio.io/istio/security/pkg/listwatch"
+	"istio.io/istio/security/pkg/pki/ca"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+)
+
+/* #nosec: disable gas linter */
+const (
+	// The Istio secret annotation type
+	IstioSecretType = "istio.io/key-and-cert"
+
+	// The ID/name for the certificate chain file.
+	CertChainID = "cert-chain.pem"
+	// The ID/name for the private key file.
+	PrivateKeyID = "key.pem"
+	// The ID/name for the CA root certificate file.
+	RootCertID = "root-cert.pem"
+	// The key to specify corresponding service account in the annotation of K8s secrets.
+	ServiceAccountNameAnnotationKey = "istio.io/service-account.name"
+
+	secretNamePrefix = "istio."
+	// For debugging, set the resync period to be a shorter period.
+	secretResyncPeriod = 10 * time.Second
+	// secretResyncPeriod = time.Minute
+
+	recommendedMinGracePeriodRatio = 0.2
+	recommendedMaxGracePeriodRatio = 0.8
+
+	// The size of a private key for a leaf certificate.
+	keySize = 2048
+
+	// The number of retries when requesting to create secret.
+	secretCreationRetry = 3
+
+	// The interval for reading a certificate
+	certReadInterval = 500 * time.Millisecond
+	// The number of tries for reading a certificate
+	maxNumCertRead = 20
+
+	// The path storing the CA certificate of the k8s apiserver
+	caCertPath = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+	// The delay introduced to debounce the CA cert events
+	watchDebounceDelay = 100 * time.Millisecond
+)
+
+type NetStatus int
+
+const (
+	Reachable NetStatus = iota
+	UnReachable
+)
+
+var (
+	// Service accounts for the webhooks.
+	WebhookServiceAccounts = []string{
+		// This is a service account created as a demo of the prototype
+		"istio-protomutate-service-account",
```

Comment:
Why do you need to work on a fake service account? Any problem using the service accounts of the webhooks?

Reply:
The service account here is for running the prototype on a demo webhook. There is no problem using the service account of the webhooks. After the demo, this service account will be replaced by the service accounts of the webhooks. I have added a TO-DO item for the replacement.

Reply:
Let's just use the service accounts for the webhooks. I don't see any risk here, since Chiron won't be deployed by default.

Reply:
Done: removed service accounts.